pkg/graveler/graveler_test.go

```diff
@@ -352,6 +356,143 @@ func TestGraveler_Set(t *testing.T) {
 		})
 	}
 }
+
+func TestGravelerSet_Advanced(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	ctx := context.Background()
+	newSetVal := graveler.ValueRecord{Key: []byte("key"), Value: &graveler.Value{Data: []byte("newValue"), Identity: []byte("newIdentity")}}
+	// RefManager mock base setup
+	refMgr := mock.NewMockRefManager(ctrl)
+	var getFlow = func() {
+		refMgr.EXPECT().ParseRef(gomock.Any()).Times(1).Return(graveler.RawRef{BaseRef: ""}, nil)
+		refMgr.EXPECT().ResolveRawRef(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(&graveler.ResolvedRef{
+			Type:                   0,
+			ResolvedBranchModifier: 0,
+			BranchRecord: graveler.BranchRecord{
+				BranchID: "",
+				Branch: &graveler.Branch{
+					CommitID:     "",
+					StagingToken: "",
+					SealedTokens: nil,
+				},
+			},
+		}, nil)
+		refMgr.EXPECT().GetCommit(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(nil, graveler.ErrCommitNotFound)
+	}
+
+	tests := []struct {
+		name                string
+		expectedValueResult graveler.ValueRecord
+		expectedErr         error
+		committedMgr        *testutil.CommittedFake
+		stagingMgr          *testutil.StagingFake
+		refSetup            func()
+	}{
+		{
+			name:         "update failure",
+			committedMgr: &testutil.CommittedFake{},
+			stagingMgr: &testutil.StagingFake{
+				UpdateErr: ErrGravelerUpdate,
+			},
+			refSetup: func() {
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(&graveler.Branch{}, nil)
+				getFlow()
+			},
+			expectedErr: ErrGravelerUpdate,
+		},
+		{
+			name:         "branch deleted after update",
+			committedMgr: &testutil.CommittedFake{},
+			stagingMgr:   &testutil.StagingFake{},
+			refSetup: func() {
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(&graveler.Branch{}, nil)
+				getFlow()
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(nil, graveler.ErrNotFound)
+			},
+			expectedErr: graveler.ErrNotFound,
+		},
+		{
+			name:         "branch token changed after update - one retry",
+			committedMgr: &testutil.CommittedFake{},
+			stagingMgr:   &testutil.StagingFake{},
+			refSetup: func() {
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(&graveler.Branch{}, nil)
+				getFlow()
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(2).Return(&graveler.Branch{
+					StagingToken: "new_token",
+				}, nil)
+				getFlow()
+				refMgr.EXPECT().GetBranch(gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(&graveler.Branch{
+					StagingToken: "new_token",
+				}, nil)
+			},
+			expectedValueResult: newSetVal,
+			expectedErr:         nil,
+		},
+		{
+			name:         "branch token changed max retries",
+			committedMgr: &testutil.CommittedFake{},
+			stagingMgr:   &testutil.StagingFake{},
+			refSetup: func() {
+				const maxRetries = 3
```

Comment:
How about making [`setTries`](https://github.com/treeverse/lakeFS/blob/5dd9fddd0b664dac4c632bc40b215dc322760921/pkg/graveler/graveler.go#L1408) public and use it here. Will prevent failure in case of future changes

Maybe we should make the `setTries` const (define in `graveler` `safeBranchWrite` in `graveler`